### PR TITLE
Switch to specular head info fetching

### PR DESCRIPTION
**Pull request by LEAFERx**
# Core changes

- Implemented specular head info fetching to correctly get finalized epoch info when magi starts

### CODE REVIEW

File: src/derive/stages/attributes.rs [hunk 1]
```diff
@@ -323,8 +323,8 @@ impl TryFrom<Log> for UserDeposited {
             .into_bytes()
             .unwrap();
 
-        let from = Address::try_from(log.topics[1])?;
-        let to = Address::try_from(log.topics[2])?;
+        let from = Address::from(log.topics[1]);
+        let to = Address::from(log.topics[2]);
```
Comment (ujvl) [inline src/derive/stages/attributes.rs:327]: why this change, is this guaranteed to succeed

Comment (LEAFERx) [reply src/derive/stages/attributes.rs:327]: yes see https://github.com/a16z/magi/actions/runs/7011622482/job/19074567685

File: src/driver/mod.rs [hunk 2]
```diff
@@ -75,9 +75,16 @@ impl Driver<EngineApi> {
...
             info::HeadInfoQuery::get_head_info(&info::HeadInfoFetcher::from(&provider), &config)
-                .await;
+                .await
+        } else {
+            specular::info::HeadInfoQuery::get_head_info(
+                &specular::info::HeadInfoFetcher::from(&provider),
+                &config,
+            )
+            .await
+        };
```
Comment (ujvl) [inline src/driver/mod.rs:87]: how about we just pass `head` in?

File: src/specular/info.rs [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        if let Some(block) = p
+            .get_block_with_txs(BlockId::Number(BlockNumber::Finalized))
+            .await
+            .ok()
+            .flatten()
+        {
+            if let Ok(info) = to_specular_head_info(p, block).await {
+                return info;
+            }
+        }
```
Comment (ujvl) [inline src/specular/info.rs:72]: can refactor this maybe

**Review by ujvl — APPROVED** [review]
